### ISSUE TITLE
Is there a way to put in the cache an http header?

### ISSUE THREAD

**Issue by IOhacker** [Trendyol/Jdempotent#24]
Hello,

Is there a way to put in the cache an http header? like

x-idempotency-key=UUID-GENERATED 

Or is there a way to put only a key/value of the Json body request?

Regards

**Comment by IOhacker** [comment]
I have fixed this using another Annotation, let me prepare a PR if it is not suitable, at least can show an approach how can I made it work for the scenario like the one described here

https://docs.google.com/presentation/d/13lw2JqE9ujHVFwmynjJcVIUA3NfcDkbs729VLgadIXE/edit?usp=sharing

**Comment by IOhacker** [comment]
And it works with this CURL

curl --location --request POST 'http://localhost:8080/send-email' \
--header 'x-idempotency-key: f3d8728e-f0f4-4969-8481-78de9d90caa1' \
--header 'Content-Type: application/json' \
--data-raw '{
        "email": "victor.romero@XXXXXXX.mx",
        "subject": "IDEMPOTENCY",
        "message": "TEST"
}'

**Comment by memojja** [comment]
@JdempotentId future in live but maven central sync is time consuming process.
I hope max in 10 hour, it wil be sync Jdempotent v1.0.8 


https://mvnrepository.com/artifact/com.trendyol/Jdempotent-spring-boot-redis-starter

**Comment by IOhacker** [comment]
I will try it. :)

**Comment by IOhacker** [comment]
I have a query, why now the annotations have been deprecated?

**Comment by memojja** [comment]
actually we will change a little typo not fundamantally deprecated.

for example:

it will be `@JdempotentResource` instead of `@IdempotentResource`

**Comment by IOhacker** [comment]
I have tried with this piece of code:


   @POST
    @Consumes({ MediaType.APPLICATION_JSON })
    @Produces({ MediaType.APPLICATION_JSON })
    @Operation(summary = "Create a Savings Product", description = "Creates a Savings Product\n\n"
            + "Mandatory Fields: name, shortName, description, currencyCode, digitsAfterDecimal,inMultiplesOf, nominalAnnualInterestRate, interestCompoundingPeriodType, interestCalculationType, interestCalculationDaysInYearType,accountingRule\n\n"
            + "Mandatory Fields for Cash based accounting (accountingRule = 2): savingsReferenceAccountId, savingsControlAccountId, interestOnSavingsAccountId, incomeFromFeeAccountId, transfersInSuspenseAccountId, incomeFromPenaltyAccountId\n\n"
            + "Optional Fields: minRequiredOpeningBalance, lockinPeriodFrequency, lockinPeriodFrequencyType, withdrawalFeeForTransfers, paymentChannelToFundSourceMappings, feeToIncomeAccountMappings, penaltyToIncomeAccountMappings, charges, allowOverdraft, overdraftLimit, minBalanceForInterestCalculation,withHoldTax,taxGroupId,accountMapping")
    @RequestBody(required = true, content = @Content(schema = @Schema(implementation = SavingsProductsApiResourceSwagger.PostSavingsProductsRequest.class)))
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "OK", content = @Content(schema = @Schema(implementation = SavingsProductsApiResourceSwagger.PostSavingsProductsResponse.class))) })
    @IdempotentResource(ttl = 30, ttlTimeUnit = TimeUnit.SECONDS)
    public String create(@IdempotentRequestPayload @HeaderParam("x-idempotency-key") String xIdempotencyKeyHeader,
            @Parameter(hidden = true) final String apiRequestBodyAsJson) {

        final CommandWrapper commandRequest = new CommandWrapperBuilder().createSavingProduct().withJson(apiRequestBodyAsJson).build();

        final CommandProcessingResult result = this.commandsSourceWritePlatformService.logCommandSource(commandRequest);

        return this.toApiJsonSerializer.serialize(result);
    }

But I got different keys

fineract-server_1  | 2021-11-02 18:42:30.763  INFO 1 --- [           main] org.apache.fineract.ServerApplication    : Started ServerApplication in 90.384 seconds (JVM running for 91.264)
fineract-server_1  | 2021-11-02 18:43:18.921 DEBUG 1 --- [nio-8443-exec-6] c.t.j.core.aspect.IdempotentAspect       : SavingsProductsApiResource.create() starting for IdempotentRequestWrapper [request=IdempotentIgnorableWrapper{nonIgnoredFields={serialPersistentFields=[Ljava.io.ObjectStreamField;@161c777d, CASE_INSENSITIVE_ORDER=java.lang.String$CaseInsensitiveComparator@7c2990fc, coder=0, serialVersionUID=-6849794470754667710, LATIN1=0, UTF16=1, COMPACT_STRINGS=true, value=[B@a0c3ee1, hash=0}}]
fineract-server_1  | 2021-11-02 18:43:19.418 DEBUG 1 --- [nio-8443-exec-6] c.t.j.core.aspect.IdempotentAspect       : SavingsProductsApiResource.create() saved to cache with IdempotencyKey [keyValue=23d82f94d6d88a8999d713325392f0]
fineract-server_1  | 2021-11-02 18:43:19.566 DEBUG 1 --- [nio-8443-exec-6] c.t.j.core.aspect.IdempotentAspect       : SavingsProductsApiResource.create() ended for IdempotentRequestWrapper [request=IdempotentIgnorableWrapper{nonIgnoredFields={serialPersistentFields=[Ljava.io.ObjectStreamField;@161c777d, CASE_INSENSITIVE_ORDER=java.lang.String$CaseInsensitiveComparator@7c2990fc, coder=0, serialVersionUID=-6849794470754667710, LATIN1=0, UTF16=1, COMPACT_STRINGS=true, value=[B@a0c3ee1, hash=0}}]
fineract-server_1  | 2021-11-02 18:43:40.436 DEBUG 1 --- [nio-8443-exec-7] c.t.j.core.aspect.IdempotentAspect       : SavingsProductsApiResource.create() starting for IdempotentRequestWrapper [request=IdempotentIgnorableWrapper{nonIgnoredFields={serialPersistentFields=[Ljava.io.ObjectStreamField;@161c777d, CASE_INSENSITIVE_ORDER=java.lang.String$CaseInsensitiveComparator@7c2990fc, coder=0, serialVersionUID=-6849794470754667710, LATIN1=0, UTF16=1, COMPACT_STRINGS=true, value=[B@5fccf707, hash=0}}]
fineract-server_1  | 2021-11-02 18:43:40.438 DEBUG 1 --- [nio-8443-exec-7] c.t.j.core.aspect.IdempotentAspect       : SavingsProductsApiResource.create() saved to cache with IdempotencyKey [keyValue=4a8af9f4c4e0fdeb6004fe514fc39f3]
fineract-server_1  | 2021-11-02 18:43:40.448  WARN 1 --- [nio-8443-exec-7] o.d.jdbc.internal.mysql.MySQLProtocol    : Could not execute query org.drizzle.jdbc.internal.common.query.DrizzleParameterizedQuery@659b63da: Duplicate entry 'DIT4' for key 'sp_unq_name'
fineract-server_1  | 2021-11-02 18:43:40.471 DEBUG 1 --- [nio-8443-exec-7] c.t.j.core.aspect.IdempotentAspect       : SavingsProductsApiResource.create() deleted from cache with IdempotencyKey [keyValue=4a8af9f4c4e0fdeb6004fe514fc39f3] . Exception : {}
fineract-server_1  | 
fineract-server_1  | org.apache.fineract.infrastructure.core.exception.PlatformDataIntegrityException: Savings product with name `DIT4` already exists
fineract-server_1  | 	at org.apache.fineract.portfolio.savings.service.SavingsProductWritePlatformServiceJpaRepositoryImpl.handleDataIntegrityIssues(SavingsProductWritePlatformServiceJpaRepositoryImpl.java:91)

**Comment by memojja** [comment]
Hello,

interesting, are you sure you wrote the same header value? 
Can you send the curl request you tried twice?

I will add that case as a example project and i will send that project

**Comment by memojja** [comment]
you are right i think, has a bug. i will fix that. thank you. I'll let you know.

**Comment by memojja** [comment]
![image](https://user-images.githubusercontent.com/2750465/139942550-ff6a7b87-4a56-4c69-a951-892aa772081c.png)

**Comment by IOhacker** [comment]
Hello Mehmet! It is working as expected!!!

Full log available at https://paste.apache.org/b145d 

1. I have tried with a simple String "hola" - It works caching the Idempotent Key
2. I have tried with a UUID String - with the same body content and sends the exception because the existing object will send and exeption and it is working not caching the Idempotent Key.
3. I have tried with the UUID String (the same as the previous test) now with different body and it works caching the Idempotent Key

I will send this proposal to the Apache Fineract team and if it is approved, I will do the proper changes for having the Jdempotent implemented in the project.

I have raised another ticket about the deprecated annotation, this is because, the strict compilation process doesn't allow any deprecated method as part of the Apache Fineract project.

Thank you!!!
